osu.Game/Scoring/Legacy/LegacyScoreEncoder.cs

```diff
@@ -27,15 +31,24 @@ public class LegacyScoreEncoder
         public const int FIRST_LAZER_VERSION = 30000000;
 
         private readonly Score score;
-        private readonly IBeatmap beatmap;
+        private readonly IBeatmap? beatmap;
 
-        public LegacyScoreEncoder(Score score, IBeatmap beatmap)
+        /// <summary>
+        /// Create a new score encoder for a specific score.
+        /// </summary>
+        /// <param name="score">The score to be encoded.</param>
+        /// <param name="beatmap">The beatmap used to convert frames for the score. May be null if the frames are already <see cref="LegacyReplayFrame"/>s.</param>
+        /// <exception cref="ArgumentException"></exception>
+        public LegacyScoreEncoder(Score score, IBeatmap? beatmap)
         {
             this.score = score;
             this.beatmap = beatmap;
 
-            if (score.ScoreInfo.BeatmapInfo.Ruleset.OnlineID < 0 || score.ScoreInfo.BeatmapInfo.Ruleset.OnlineID > 3)
-                throw new ArgumentException("Only scores in the osu, taiko, catch, or mania rulesets can be encoded to the legacy score format.", nameof(score));
+            if (beatmap == null && !score.Replay.Frames.All(f => f is LegacyReplayFrame))
+                throw new ArgumentException(@"Beatmap must be provided if frames are not already legacy frames.", nameof(beatmap));
+
+            if (score.ScoreInfo.Ruleset.OnlineID < 0 || score.ScoreInfo.Ruleset.OnlineID > 3)
```

Comment:
nit: this was here before, but the `3` can be replaced with `ILegacyRuleset.MAX_LEGACY_RULESET_ID`